src/engine/thread.js

```diff
@@ -321,14 +340,22 @@ class Thread {
         this.justReported = typeof value === 'undefined' ? null : value;
     }
 
+    /**
+     * Return a execution context for a block to use.
+     * @returns {object} the execution context
+     */
+    getExecutionContext () {
+        const frame = this.stackFrame;
+        frame.needReset = true;
+        return frame.executionContext;
+    }
+
     /**
      * Initialize procedure parameters on this stack frame.
      */
     initParams () {
-        const stackFrame = this.peekStackFrame();
-        if (stackFrame.params === null) {
-            stackFrame.params = {};
-        }
+        const stackFrame = this.stackFrame;
+        stackFrame.params = Object.create(null);
```

Comment:
does it make sense to just use `defaultParams` here?

Reply:
Do we want to guard against overwriting already existing params here?

Reply:
Ooop. That is what I meant to do. :smile:

Reply:
Hmm. Aw no, I think I misunderstood when I last responded.

This is compatible with Scratch 2 procedure behaviour to my knowledge. This PR swaps the order the stack is changed. In this PR the stack is pushed and then parameters are set. The parent procedure's stack frame is left alone and the new stack frame will start a new set of parameters.

Reply:
Ah okay.

That still leaves the question of whether we should use `defaultParams` here instead of explicitly using `Object.create(null)` again.